server/src/main/java/de/uftos/utils/LessonsDeleter.java

```diff
@@ -0,0 +1,110 @@
+public class LessonsDeleter {
+
+  private final LessonRepository lessonRepository;
+  private final TimetableRepository timetableRepository;
+
+  /**
+   * Creates the lesson deleter.
+   *
+   * @param lessonRepository    the repository for accessing the lesson table.
+   * @param timetableRepository the repository for accessing the timetable table.
+   */
+  public LessonsDeleter(LessonRepository lessonRepository,
+                        TimetableRepository timetableRepository) {
+    this.lessonRepository = lessonRepository;
+    this.timetableRepository = timetableRepository;
+  }
+
+  /**
+   * Deletes all lessons that use these teachers and all timetables tha
+   * use these lessons.
+   *
+   * @param teachers the teachers for which the lessons have to be deleted.
+   */
+  public void fromTeachers(List<Teacher> teachers) {
+    List<Lesson> lessons = this.lessonRepository.findAllByTeacher(teachers);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+  /**
+   * Deletes all lessons that use these subjects and all timetables tha
+   * use these lessons.
+   *
+   * @param subjects the subjects for which the lessons have to be deleted.
+   */
+  public void fromSubjects(List<Subject> subjects) {
+    List<Lesson> lessons = this.lessonRepository.findAllBySubject(subjects);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+  /**
+   * Deletes all lessons that use these rooms and all timetables tha
+   * use these lessons.
+   *
+   * @param rooms the rooms for which the lessons have to be deleted.
+   */
+  public void fromRooms(List<Room> rooms) {
+    List<Lesson> lessons = this.lessonRepository.findAllByRoom(rooms);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+
+  /**
+   * Deletes all lessons that use these student groups and all timetables tha
+   * use these lessons.
+   *
+   * @param studentGroups the studentGroups for which the lessons have to be deleted.
+   */
+  public void fromStudentGroups(List<StudentGroup> studentGroups) {
+    List<Lesson> lessons = this.lessonRepository.findAllByStudentGroup(studentGroups);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+  /**
+   * Deletes all lessons that use these timeslots and all timetables tha
+   * use these lessons.
+   *
+   * @param timeslots the timeslots for which the lessons have to be deleted.
+   */
+  public void fromTimeSlots(List<Timeslot> timeslots) {
+    List<Lesson> lessons = this.lessonRepository.findAllByTimeslot(timeslots);
+    deleteLessonsAndTimetables(lessons);
+  }
+
+  private void deleteLessonsAndTimetables(List<Lesson> lessons) {
+    if (lessons.isEmpty()) {
+      return;
+    }
+    List<Timetable> timetables = this.timetableRepository.findAllByLessons(
+        lessons.stream().map(Lesson::getId).toList()
+    );
+
+    // Adding all lessons from the timetables to be deleted
+    lessons.addAll(
+        timetables
+            .stream()
+            .map(Timetable::getLessons)
+            .flatMap(List::stream)
+            .toList()
+    );
```

Comment:
This will result in duplicate entries in the List, but I guess the deletion will still handle those correctly.
```Java
lessons = timetables.stream().[...]
``` 
should also work

Reply:
I replace the my code with your suggestion and I am now using an HashSet to avoid duplicated, even though it shouldn't have caused any problems.